handler/plchandler/plchandler.go

```diff
@@ -54,84 +71,125 @@ func NewPLCHandler(core handler.CoreHandler) *PLCHandler {
+	for _, mb := range tagsMemBlocks {
+		if block.size == 0 {
+			block.address = mb.address
+			block.size = mb.size
+			block.tags = append(block.tags, mb.tags[0])
+			continue
+		}
+		newBlockSize := mb.address + mb.size - block.address
+		breakSize := mb.address - block.address - block.size
+		if newBlockSize > maxBlockSize || breakSize > maxBreakSize {
+			rv = append(rv, block)
+			block = MemBlock{address: mb.address, size: mb.size}
+			block.tags = append(block.tags, mb.tags[0])
+			continue
+		}
+		block.size = newBlockSize
+		block.tags = append(block.tags, mb.tags[0])
+	}
+	if block.size > 0 {
+		rv = append(rv, block)
+	}
+
+	return rv
+}
+
 func (ph *PLCHandler) Start() error {
 	ph.BaseHandler.Start()
 
-	ph.Values = make(map[string]string)
+	ph.tags = make(map[string]*Tag)
 
 	tags := getTagsConfig()
+	tagsCount := len(tags.Input) + len(tags.Output)
+	tagsMemBlocks := make([]MemBlock, 0, tagsCount)
 	for _, tag := range tags.Input {
-		ph.Tags = append(ph.Tags, tag)
+		tag.Size = 1
+		tag.Type = typeInput
+		ph.tags[tag.Name] = &tag
+		mb := MemBlock{address: tag.Address, size: tag.Size, tags: []*Tag{&tag}}
+		tagsMemBlocks = append(tagsMemBlocks, mb)
 	}
 	for _, tag := range tags.Output {
-		ph.Tags = append(ph.Tags, tag)
+		tag.Size = 1
+		tag.Type = typeOutput
+		ph.tags[tag.Name] = &tag
+		mb := MemBlock{address: tag.Address, size: tag.Size, tags: []*Tag{&tag}}
+		tagsMemBlocks = append(tagsMemBlocks, mb)
 	}
-	fmt.Printf("%+v\n", ph.Tags)
+	ph.pollingMemBlocks = makePollingMemBlocks(tagsMemBlocks)
 
-	err := ph.ClientHandler.Connect()
+	err := ph.clientHandler.Connect()
 	if err != nil {
 		return err
 	}
-	defer ph.ClientHandler.Close()
-	//
-	client := modbus.NewClient(ph.ClientHandler)
+	defer ph.clientHandler.Close()
+
+	client := modbus.NewClient(ph.clientHandler)
+	return ph.loop(&client)
+}
+
+func (ph *PLCHandler) loop(client *modbus.Client) error {
 	for {
 		select {
 		case <-ph.Ctx.Done():
 			fmt.Println("Context canceled")
 			return ph.Ctx.Err()
 		case command := <-ph.CommandChanIn:
-			// fmt.Printf("have command: %v\n", command)
-			for _, tag := range ph.Tags {
-				if tag.Name == command.Name {
-					// fmt.Printf("address: %v\n", tag.Address)
-					var val uint16 = ON
-					if command.Value == "0" {
-						val = OFF
-					}
-					_, err = client.WriteSingleCoil(uint16(tag.Address), val)
-					if err != nil {
-						fmt.Printf("error: %v\n", err)
-					}
+			//fmt.Printf("have command: %v\n", command)
+			tag, ok := ph.tags[command.Name]
+			if ok {
+				var val uint16 = on
+				if command.Value == "0" {
+					val = off
+				}
+				_, err := (*client).WriteSingleCoil(uint16(tag.Address), val)
+				if err != nil {
+					fmt.Printf("error: %v\n", err)
 				}
 			}
 		default:
-			for _, tag := range ph.Tags {
-				results, err := client.ReadCoils(uint16(tag.Address), 1)
+			for _, mb := range ph.pollingMemBlocks {
+				results, err := (*client).ReadCoils(uint16(mb.address), uint16(mb.size))
```

Comment:
no need for all uint16 conversation if we change address and size to uint16 in a first place

Reply:
fixed